specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json

```diff
@@ -0,0 +1,3268 @@
+        "tags": [
+          "Cluster"
+        ],
+        "x-ms-examples": {
+          "List cluster by resource group": {
+            "$ref": "./examples/ClusterListByResourceGroupOperation_example.json"
+          }
+        },
+        "responses": {
+          "200": {
+            "description": "OK - Get cluster  successfully",
+            "schema": {
+              "$ref": "#/definitions/ClusterListResult"
+            }
+          },
+          "default": {
+            "description": "Error",
+            "schema": {
+              "$ref": "#/definitions/ErrorModel"
+            }
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.ServiceFabric/clusters": {
+      "get": {
+        "operationId": "Clusters_List",
+        "summary": "List cluster resource",
+        "description": "List cluster resource\n",
+        "parameters": [
+          {
+            "$ref": "#/parameters/api-version"
+          },
+          {
+            "$ref": "#/parameters/subscriptionId"
+          }
+        ],
+        "tags": [
+          "Cluster"
+        ],
+        "x-ms-examples": {
+          "List clusters": {
+            "$ref": "./examples/ClusterListOperation_example.json"
+          }
+        },
+        "responses": {
+          "200": {
+            "description": "OK - Get cluster  successfully",
+            "schema": {
+              "$ref": "#/definitions/ClusterListResult"
+            }
+          },
+          "default": {
+            "description": "Error",
+            "schema": {
+              "$ref": "#/definitions/ErrorModel"
+            }
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.ServiceFabric/locations/{location}/clusterVersions/{clusterVersion}": {
+      "get": {
+        "operationId": "ClusterVersions_Get",
+        "summary": "Get cluster code versions",
+        "description": "Get cluster code versions by location\n",
+        "parameters": [
+          {
+            "$ref": "#/parameters/locationForClusterCodeVersions"
+          },
+          {
+            "$ref": "#/parameters/api-version"
+          },
+          {
+            "$ref": "#/parameters/subscriptionId"
+          },
+          {
+            "$ref": "#/parameters/clusterVersion"
+          }
+        ],
+        "x-ms-examples": {
+          "Get cluster version": {
+            "$ref": "examples/ClusterVersionsGet_example.json"
+          }
+        },
+        "tags": [
+          "ClusterVersion"
+        ],
+        "responses": {
+          "200": {
+            "description": "OK - Get cluster code versions successfully",
+            "schema": {
+              "$ref": "#/definitions/ClusterCodeVersionsListResult"
+            }
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.ServiceFabric/locations/{location}/environments/{environment}/clusterVersions/{clusterVersion}": {

```

Comment:
How does customer discover the list of environments?

Reply:
environments is a property on the response of clusterVersions. This could be supported using a query parameter instead of modeling it as a resource. So to get cluster versions for a specific environment, customer will use this API -
/subscriptions/{subscriptionId}/providers/Microsoft.ServiceFabric/locations/{location}/clusterVersions and then do a $filter.